yolact_ros2_3d/include/yolact_ros2_3d/YolactROS23D.hpp

```diff
@@ -15,22 +15,90 @@
 /* Author: Francisco Martín fmrico@gmail.com */
 /* Author: Fernando González fergonzaramos@yahoo.es */
 
+#ifndef YOLACT_ROS2_3D__YOLACTROS23D_HPP_
+#define YOLACT_ROS2_3D__YOLACTROS23D_HPP_
+
+#include <tf2/convert.h>
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/transform_listener.h>
+#include <cv_bridge/cv_bridge.h>
 #include <rclcpp/rclcpp.hpp>
 #include <rclcpp_lifecycle/lifecycle_node.hpp>
-
-#ifndef YOLACT_ROS2_3D__YOLACT_ROS2_3D_HPP__
-#define YOLACT_ROS2_3D__YOLACT_ROS2_3D_HPP__
+#include <lifecycle_msgs/msg/state.hpp>
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <sensor_msgs/msg/point_cloud.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/core/mat.hpp>
+#include <visualization_msgs/msg/marker_array.hpp>
+#include <string>
+#include <vector>
+#include <map>
+#include "yolact_ros2_msgs/msg/detections.hpp"
+#include "yolact_ros2_msgs/msg/detection.hpp"
+#include "gb_visual_detection_3d_msgs/msg/bounding_boxes3d.hpp"
+#include "gb_visual_detection_3d_msgs/msg/bounding_box3d.hpp"
 
 namespace yolact_ros2_3d
 {
 class YolactROS23D : public rclcpp_lifecycle::LifecycleNode
 {
 public:
-	YolactROS23D();
+  YolactROS23D();
+
+  void update();
+
 private:
+  using CallbackReturnT =
+    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
+
+  CallbackReturnT on_configure(const rclcpp_lifecycle::State & state);
+  CallbackReturnT on_activate(const rclcpp_lifecycle::State & state);
+  CallbackReturnT on_deactivate(const rclcpp_lifecycle::State & state);
+  CallbackReturnT on_cleanup(const rclcpp_lifecycle::State & state);
+  CallbackReturnT on_shutdown(const rclcpp_lifecycle::State & state);
+  CallbackReturnT on_error(const rclcpp_lifecycle::State & state);
+
+  /* ----------- */
+
+  void pointCloudCb(const sensor_msgs::msg::PointCloud2::SharedPtr msg);
+  void yolactCb(const yolact_ros2_msgs::msg::Detections::SharedPtr msg);
+  void calculate_boxes(
+    sensor_msgs::msg::PointCloud2 cloud_pc2, sensor_msgs::msg::PointCloud cloud_pc,
+    gb_visual_detection_3d_msgs::msg::BoundingBoxes3d * boxes);
+  void getMask(yolact_ros2_msgs::msg::Detection det, cv::Mat * output_mask);
+  void erodeMask(std::string class_name, cv::Mat * mask, cv::Mat * eroded_mask);
+  void publishMarkers(gb_visual_detection_3d_msgs::msg::BoundingBoxes3d boxes);
+
+  bool setErodingFactors();
+  bool pixelBelongsToBbox(const yolact_ros2_msgs::msg::Mask & mask, size_t x, size_t y);
+  bool calculateBbox(
```

Comment:
You are passing these parameters by value, which is time costly. 

Change `sensor_msgs::msg::PointCloud2 cloud2` to:
- `sensor_msgs::msg::PointCloud2 & cloud2`
- or `const sensor_msgs::msg::PointCloud2 & cloud2` if it is not modified inside.

Do the same in all methods and parameters for all non-basic types. You will see a significant improvement 😉

Reply:
It is observed a great improvement. Computational charge decreases significantly.